volatility/framework/plugins/windows/dumpfiles.py

```diff
@@ -0,0 +1,245 @@
+# This file is Copyright 2019 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import logging
+import ntpath
+from volatility.framework import interfaces, renderers, exceptions, constants
+from volatility.plugins.windows import handles
+from volatility.plugins.windows import pslist
+from volatility.framework.configuration import requirements
+from volatility.framework.renderers import format_hints
+from volatility.framework.objects import utility
+from typing import List, Tuple
+vollog = logging.getLogger(__name__)
+
+FILE_DEVICE_DISK = 0x7
+FILE_DEVICE_NETWORK_FILE_SYSTEM = 0x14
+EXTENSION_CACHE_MAP = {
+    "dat": "DataSectionObject",
+    "img": "ImageSectionObject",
+    "vacb": "SharedCacheMap",
+}
+
+class DumpFiles(interfaces.plugins.PluginInterface):
+    """Dumps cached file contents from Windows memory samples."""
+
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        # Since we're calling the plugin, make sure we have the plugin's requirements
+        return [
+            requirements.TranslationLayerRequirement(name='primary',
+                                                     description='Memory layer for the kernel',
+                                                     architectures=["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name="nt_symbols", description="Windows kernel symbols"),
+            requirements.IntRequirement(name='pid',
+                                        description="Process ID to include (all other processes are excluded)",
+                                        optional=True),
+            requirements.IntRequirement(name='virtaddr',
+                                        description="Dump a single _FILE_OBJECT at this virtual address",
+                                        optional=True),
+            requirements.IntRequirement(name='physaddr',
+                                        description="Dump a single _FILE_OBJECT at this physical address",
+                                        optional=True),
+            requirements.PluginRequirement(name='pslist', plugin=pslist.PsList, version=(1, 0, 0)),
+            requirements.PluginRequirement(name='handles', plugin=handles.Handles, version=(1, 0, 0))
+        ]
+
+    def dump_file_producer(self, file_object: interfaces.objects.ObjectInterface,
```

Comment:
It would be better if this returned a `FileInterface` object, so it doesn't need `self` and can become a `classmethod`?  I'd return `None` in place of error text, but definitely include useful information in a `vollog.debug` message.

Reply:
Either that, or we could actually raise an Exception?  It'll need catching in either circumstance?  Even though I'm usually a fan of exceptions, they do tend to require a lot of code for catching, so as long as it's caught and logged in here, I'm ok returning None.

Reply:
Consistency would suggest adding a `context` parameter, and using `layer_name` rather than `layer` (simple variables over complex ones carrying their own context, but in this circumstance, because the context would only be used to lookup the layer, I'll let the layer be passed in whole.  Equally happy for people to consider it inconsistent and us to add the context in as a parameter too, so takes your choice.  5;P